R/fair-aaa.R

```diff
@@ -0,0 +1,181 @@
+#' Create fairness metrics
+#'
+#' Fairness metrics quantify the disparity in value of a metric across a number
+#' of groups. Fairness metrics with a value of zero indicate that the
+#' underlying metric has parity across groups. yardstick defines
+#' several common fairness metrics using this function, such as
+#' [demographic_parity()], [equal_opportunity()], and [equalized_odds()].
+#'
+#' @param .fn A yardstick metric function or metric set.
+#' @param .name The name of the metric to place in the `.metric` column
+#' of the output.
+#' @param .post A function to post-process the generated metric set results `x`.
+#' In many cases, `~diff(range(x$.estimate))` or
+#' `~r <- range(x$.estimate); r[1]/r[2]`.
+#'
+#' @section Relevant Group Level:
+#' Additional arguments can be passed to the function outputted by
+#' the function that this function outputs. That is:
+#'
+#' ```
+#' res_fairness <- fairness_metric(...)
+#' res_by <- res_fairness(by)
+#' res_by(..., additional_arguments_to_.post = TRUE)
+#' ```
+#'
+#' For finer control of how groups in `by` are treated, use the
+#' `.post` argument.
+#'
+#' @return
+#' This function is a
+#' [function factory](https://adv-r.hadley.nz/function-factories.html); it's
+#' output is itself a function. Further, the functions that this function
+#' outputs are also function factories. More explicitly, this looks like:
+#'
+#' ```
+#' # a function with similar implementation to `demographic_parity()`:
+#' diff_range <- function(x, ...) {diff(range(x$.estimate))}
+#'
+#' dem_parity <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "dem_parity",
+#'     .post = diff_range
+#'   )
+#' ```
+#'
+#' The outputted `dem_parity` is a function that takes one argument, `by`,
+#' indicating the data-masked variable giving the sensitive feature.
+#'
+#' When called with a `by` argument, `dem_parity` will return a yardstick
+#' metric function like any other:
+#'
+#' ```
+#' dem_parity_by_gender <- dem_parity(gender)
+#' ```
+#'
+#' Note that `dem_parity` doesn't take any arguments other than `by`, and thus
+#' knows nothing about the data it will be applied to other than that it ought
+#' to have a column with name `"gender"` in it.
+#'
+#' The output `dem_parity_by_gender` is a metric function that takes the
+#' same arguments as the function supplied as `fn`, in this case
+#' `detection_prevalence`. It will thus interface like any other yardstick
+#' function except that it will look for a `"gender"` column in
+#' the data it's supplied.
+#'
+#' In addition to the examples below, see the documentation on the
+#' return value of fairness metrics like [demographic_parity()],
+#' [equal_opportunity()], or [equalized_odds()] to learn more about how the
+#' output of this function can be used.
+#'
+#' @examples
+#' data(hpc_cv)
```

Comment:
If we decide to move forward with a fairness-oriented name, it'd be great if we could use some example data here that has a plausibly "sensitive" attribute. yardstick doesn't `Suggests` modeldata at the moment, which has some options. `infer::gss` would also work well here.